File: src/PythonClient/src/quixstreams/statestorages/serializers/bytevalueserializer.py
import io
import pickle
import struct

from quixstreams.statestorages.statetype import StateType
from quixstreams.statestorages.statevalue import StateValue


class ByteValueSerializer:
    # Serializer written using struct module
    # Custom protocol

    CodecId = 'b'

    @staticmethod
    def serialize(state_value: StateValue) -> bytes:
        """Serializes StateValue into byte array"""
        with io.BytesIO() as stream:
            stream.write(struct.pack('c', ByteValueSerializer.CodecId.encode()))
            stream.write(struct.pack('c', state_value.type.value.encode()))

            if state_value.type == StateType.Binary:
                if state_value.value is None:
                    stream.write(struct.pack('>i', -1))
                else:
                    stream.write(struct.pack('>i', len(state_value.value)))
                    stream.write(state_value.value)

            elif state_value.type == StateType.Object:
                if state_value.value is None:
                    stream.write(struct.pack('>i', -1))
                else:
                    value_in_bytes = pickle.dumps(state_value.value)
                    stream.write(struct.pack('>i', len(value_in_bytes)))
                    stream.write(value_in_bytes)

            elif state_value.type == StateType.Bool:
                stream.write(struct.pack('?', state_value.value))

            elif state_value.type == StateType.Double:
                stream.write(struct.pack('>d', state_value.value))

            elif state_value.type == StateType.Long:
                stream.write(struct.pack('>q', state_value.value))

            elif state_value.type == StateType.String:
                stream.write(state_value.value.encode())
            else:
                raise ValueError(f'Unsupported type {state_value.type} in Serialization')

            return stream.getvalue()

    @staticmethod
    def deserialize(data: bytes) -> StateValue:
        """Deserializes byte array into StateValue"""
        with io.BytesIO(data) as stream:
            codec_id = stream.read(1).decode()
            if codec_id != ByteValueSerializer.CodecId:
                raise ValueError('Wrong codec id')

            data_type = stream.read(1).decode()

            if data_type == StateType.Binary.value:
                data_length = struct.unpack('>i', stream.read(4))[0]
                if data_length == -1:
                    return StateValue(None)
                return StateValue(stream.read(data_length))

            elif data_type == StateType.Object.value:
                data_length = struct.unpack('>i', stream.read(4))[0]
                if data_length == -1:
                    return StateValue(None)
                return StateValue(pickle.loads(stream.read(data_length)))

            elif data_type == StateType.Bool.value:
                return StateValue(struct.unpack('?', stream.read(1))[0])

            elif data_type == StateType.Double.value:
                return StateValue(struct.unpack('>d', stream.read(8))[0])

            elif data_type == StateType.Long.value:
                return StateValue(struct.unpack('>q', stream.read(8))[0])

            elif data_type == StateType.String.value:
                return StateValue(stream.read().decode())

            else:
                raise ValueError(f'Unsupported type {data_type} in Serialization')

File: src/CsharpClient/QuixStreams.State/Serializers/ByteValueSerializer.cs
using System;
using System.IO;
using System.Linq;
using QuixStreams.State.Storage.FileStorage;

namespace QuixStreams.State.Serializers
{
    /// <summary>
    /// Serializer written using the BinaryWriter and BinaryReader
    /// Custom protocol
    /// Implements crc32 checksum to check the validity of stored data
    /// </summary>
    public class ByteValueSerializer
    {
        private const char CodecId = 'b';
        private const char TypeBool = 'b';
        private const char TypeLong = 'l';
        private const char TypeDouble = 'd';
        private const char TypeBinary = 'a';
        private const char TypeObject = 'o';
        private const char TypeString = 's';

        private const int CrcChecksumLength = 32 / 8; // CRC hash 32 bit long, so 4 bytes

        private static byte[] calculateCRC32Hash(byte[] data)
        {
            Crc32 crc32 = new Crc32();
            return crc32.ComputeHash(data);
        }

        /// <summary>
        /// Serialize State value
        /// </summary>
        /// <param name="value">Value to serialize</param>
        /// <returns>Serialized byte array</returns>
        public static byte[] Serialize(StateValue value)
        {
            using (MemoryStream stream = new MemoryStream())
            {
                using (BinaryWriter writer = new BinaryWriter(stream))
                {
                    writer.Write(CodecId);
                    switch (value.Type)
                    {
                        case StateValue.StateType.Binary:
                            {
                                byte[] data = value.BinaryValue;
                                writer.Write(TypeBinary);

                                if (data != null)
                                {
                                    writer.Write(data.Length);
                                    writer.Write(data);
                                }
                                else
                                {
                                    writer.Write(-1);
                                }
                            }
                            break;
                        case StateValue.StateType.Object:
                        {
                            byte[] data = value.BinaryValue ?? Array.Empty<byte>();
                            writer.Write(TypeObject);
                            if (data != null)
                            {
                                writer.Write(data.Length);
                                writer.Write(data);
                            }
                            else
                            {
                                writer.Write(-1);
                            }
                        }
                            break;                        
                        case StateValue.StateType.Bool:
                            {
                                bool data = value.BoolValue;
                                writer.Write(TypeBool);
                                writer.Write(data);
                            }
                            break;
                        case StateValue.StateType.Double:
                            {
                                double data = value.DoubleValue;
                                writer.Write(TypeDouble);
                                writer.Write(data);
                            }
                            break;
                        case StateValue.StateType.Long:
                            {
                                long data = value.LongValue;
                                writer.Write(TypeLong);
                                writer.Write(data);
                            }
                            break;
                        case StateValue.StateType.String:
                            {
                                string data = value.StringValue ?? ""; // null would throw arg exc
                                writer.Write(TypeString);
                                writer.Write(data);
                            }
                            break;
                        default:
                            throw new NotSupportedException($"Unsupported type {value.Type} in Serialization");
                    }
                    //compute crc32 hash to check the validity and append it at the end
                    var raw = stream.ToArray();
                    byte[] hash = calculateCRC32Hash(raw);
                    writer.Write(hash);
                }
                stream.Flush();
                return stream.ToArray();
            }
        }

        private static void ValidateChecksum(byte[] data)
        {
            //check for crc32
            if (data.Length <= CrcChecksumLength)
            {
                throw new FormatException("Malformed data file");
            }
            var newData = new byte[data.Length - CrcChecksumLength];
            Array.Copy(data, 0, newData, 0, newData.Length);
            var checksum = new byte[CrcChecksumLength];
            Array.Copy(data, newData.Length, checksum, 0, CrcChecksumLength);
            
            byte[] computedHash = calculateCRC32Hash(newData); 
            if (!computedHash.SequenceEqual(checksum))
            {
                throw new FormatException("Malformed data file");
            }

        }

        /// <summary>
        /// Deserialize State value
        /// </summary>
        /// <param name="data">Byte array to deserialize into a State value</param>
        /// <returns>Deserialized State value</returns>
        public static StateValue Deserialize(byte[] data)
        {
            if (data == null)
                return null;
            
            ValidateChecksum(data);

            using (MemoryStream stream = new MemoryStream(data))
            {
                using (BinaryReader reader = new BinaryReader(stream))
                {
                    char codecID = reader.ReadChar();
                    if(codecID != CodecId)
                    {
                        throw new NotSupportedException($"Wrong codec id");
                    }
                    char type = reader.ReadChar();
                    switch (type)
                    {
                        case TypeBinary:
                            {
                                int dataLength = reader.ReadInt32();
                                if (dataLength == -1) return new StateValue((byte[])null);
                                return new StateValue(reader.ReadBytes(dataLength));
                            }
                        case TypeObject:
                            {
                                int dataLength = reader.ReadInt32();
                                if (dataLength == -1) return new StateValue((byte[])null);
                                return new StateValue(reader.ReadBytes(dataLength), StateValue.StateType.Object);
                            }                        
                        case TypeBool:
                            {
                                return new StateValue(reader.ReadBoolean());
                            }
                        case TypeDouble:
                            {
                                return new StateValue(reader.ReadDouble());
                            }
                        case TypeLong:
                            {
                                return new StateValue(reader.ReadInt64());
                            }
                        case TypeString:
                            {
                                return new StateValue(reader.ReadString());
                            }
                        default:
                            throw new NotSupportedException($"Unsupported type {type} in Serialization");
                    }
                }
            }

        }

    }
}
